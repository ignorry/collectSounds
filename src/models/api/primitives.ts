/**
 * @typedef PageInfo
 * @prop {number} totalResults
 * @prop {number} resultsPerPage
 */
export type PageInfo = {
  totalResults: number;
  resultsPerPage: number;
}

/**
 * @typedef Thumbnail
 * @prop {string} url
 * @prop {number} width
 * @prop {number} height
 */
export type Thumbnail = {
  url: string;
  width: number;
  height: number;
}

/**
 * @typedef Thumbnails
 * @prop {Thumbnail} [default]
 * @prop {Thumbnail} [medium]
 * @prop {Thumbnail} [high]
 * @prop {Thumbnail} [standard]
 * @prop {Thumbnail} [maxres]
 */
export type Thumbnails = {
  default?: Thumbnail;
  medium?: Thumbnail;
  high?: Thumbnail;
  standard?: Thumbnail;
  maxres?: Thumbnail;
}

/**
 * @typedef Localized
 * @prop {string} title
 * @prop {string} description
 */
export type Localized = {
  title: string;
  description: string;
}