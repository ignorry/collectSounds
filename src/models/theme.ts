export enum ThemeEnum  {
  light = 'light',
  dark = 'dark'
}

// all sizes in rem
export type Theme = {
  iconSize: number,
  rangeSize: number,
  spinnerSize: number,
  contentWidth: number,
  colors: {
    main: string,
    bg?: string,
    bgSecondary?: string,
    font?: string,
    darkBg?: string,
  },
  decorative: { // in px
    borderRadius: number;
    padding: number;
  },
  gaps: {
    smallest: number,
    small: number,
    big: number,
  },
  thumbnail: {
    width: number,
    height: number,
    infoColor: string,
    infoWidth: 4,
  },
  error: {
    backgroundColor: string,
    width: number,
  },
  media: {
    small: string,
  },
  font: {
    primary: {
      size: number,
      opacity: number,
    },
    secondary: {
      size: number,
      opacity: number,
    },
  },
  durations: {
    ms300: number,
  },
  order: {
    footer: number,
    modal: number,
  },
}