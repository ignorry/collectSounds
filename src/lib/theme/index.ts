import { DefaultTheme } from "styled-components";
import { ThemeEnum } from "../../models/theme";
import { baseTheme } from "./baseTheme";

export const lightTheme: DefaultTheme = {
  ...baseTheme,
  type: ThemeEnum.light,

  colors: {
    ...baseTheme.colors,
    main: '#00bb88',
    bg: '#ffffff',
    bgSecondary: '#d8d8d8',
    font: '#000000',
    darkBg: '#d9d9d9'
  },
}

export const darkTheme: DefaultTheme = {
  ...baseTheme,
  type: ThemeEnum.dark,

  colors: {
    ...baseTheme.colors,
    bg: '#000000',
    bgSecondary: '#2a2a2a',
    font: '#ffffff',
    darkBg: '#121212'
  },
}

/**
 * get theme depending on isDark parameter
 * @param {boolean} isDark 
 * @returns {DefaultTheme}
 */
export const getTheme = ( isDark: boolean ): DefaultTheme => isDark ? darkTheme : lightTheme;