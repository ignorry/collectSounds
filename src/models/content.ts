import { Thumbnails } from "./api/primitives";

/**
 * @typedef Video
 * @prop {"video"} type - type of content. always "video" for this type
 * @prop {string} id
 * @prop {string} [audioUrl] - T-One mp3 file
 * @prop {string} publishedAt
 * @prop {string} title
 * @prop {string} description
 * @prop {string} lang - language in ISO format e.g 'en'
 * @prop {string} channelTitle
 * @prop {Thumbnails} thumbnails
 * @prop {number} duration - in ms
 * @prop {number} [passed] - passed duration in ms
 * @prop {Array<string>} [tags] - given by user
 * @prop {number} lastModified - used for synchronization
 */
export type Video = {
  type: "video";
  id: string;
  audioUrl?: string;
  publishedAt: string;
  title: string;
  description: string;
  lang: string;
  channelTitle: string;
  thumbnails: Thumbnails;
  duration: number;
  passed?: number;
  tags?: Array<string>;
  lastModified: number;
}

/**
 * @typedef Playlist
 * @prop {"video"} type - type of content. always "video" for this type
 * @prop {string} id
 * @prop {Map<string, Video>} [videos] - map of videos contained in a playlist
 * @prop {string} title
 * @prop {string} description
 * @prop {string} lang - language in ISO format e.g 'en'
 * @prop {string} channelTitle
 * @prop {Thumbnails} thumbnails
 * @prop {Array<string>} [tags] - given by user
 * @prop {number} lastModified - used for synchronization
 */
export type Playlist = {
  type: "playlist";
  id: string;
  videos?: Array<{ id: string, item: Video }>;
  title: string;
  description: string;
  lang: string;
  channelTitle: string;
  thumbnails: Thumbnails;
  tags?: Array<string>;
  lastModified: number;
}

/**
 * @typedef Channel
 * @prop {"video"} type - type of content. always "video" for this type
 * @prop {string} id
 * @prop {Map<string, Video>} [videos] - map of videos contained in a channel
 * @prop {Map<string, Playlist>} [playlists] - map of videos contained in a channel
 * @prop {string} title
 * @prop {string} description
 * @prop {string} customUrl
 * @prop {Thumbnails} thumbnails
 * @prop {Array<string>} [tags] - given by user
 * @prop {number} lastModified - used for synchronization
 */
export type Channel = {
  type: "channel";
  id: string;
  videos?: Array<{ id: string, item: Video }>;
  playlists?: Array<{ id: string, item: Playlist }>;
  title: string;
  description: string;
  customUrl: string;
  thumbnails: Thumbnails;
  tags?: Array<string>;
  lastModified: number;
}

/**
 * combine content types
 */
export type Content = Video | Playlist | Channel;