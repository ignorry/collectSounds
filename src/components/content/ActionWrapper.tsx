import React, { useEffect, useRef, useState } from "react";
import styled from "styled-components";
import { Video as VideoType } from "../../models/content";
import { getDurationFromMs } from "../../lib/getDuration";

import Label from "../primitives/Label";

const Wrapper = styled.div`
  position: relative;
`;

const Left = styled.div`
  display: flex;
  align-items: center;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  color: ${ ({ theme }) => theme.colors.main };
  opacity: 0;
`;

const Right = styled.div`
  display: flex;
  align-items: center;
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  color: ${ ({ theme }) => theme.colors.main };
  opacity: 0;
`;

const Component = styled.div`
  position: relative;
  top: 0;
  bottom: 0;
`;

/**
 * @typedef Action
 * @prop {string} label
 * @prop {Function} callback
 */
export type Action = {
  label: string,
  callback: Function,
}

/**
 * @typedef Props
 * @prop {any} [children] - to be wrapped
 * @prop {Action} [left]
 * @prop {Action} [right]
 * @prop {Function} [callback] - click callback
 */
export type Props = {
  children?: any,
  left?: Action,
  right?: Action,
  callback?: Function,
};

const ActionWrapper: React.FC<Props> = ( props: Props ) => {
  const left: React.MutableRefObject<undefined | HTMLDivElement> = useRef();
  const right: React.MutableRefObject<undefined | HTMLDivElement> = useRef();
  const component: React.MutableRefObject<undefined | HTMLDivElement> = useRef();

  useEffect(() => {
    const reset = () => {
      component.current.setAttribute( 'x', '0' );
      component.current.setAttribute( 'y', '0' );
      component.current.style.left = '0';
      if ( right && right.current )
        right.current.style.opacity = '0';
      if ( left && left.current )
        left.current.style.opacity = '0';
    }

    const onStart = ( clientX: number, clientY: number ) => {
      component.current.setAttribute( 'x', clientX.toString() );
      component.current.setAttribute( 'y', clientY.toString() );
    }

    const onMove = ( clientX: number, clientY: number ) => {
      const start = +component.current.getAttribute( 'x' );
      const startY = +component.current.getAttribute( 'y' );

      if ( Math.abs( clientY-startY ) > Math.abs( clientX-start ) ) return;

      if ( ! start ) return;
      if ( clientX - start > 100 && props.left )
        component.current.style.left = `${ 100+Math.log( clientX - start - 100 )*5 }px`;
      else if ( clientX - start < -100 && props.right )
        component.current.style.left = `${ -Math.log( -( clientX - start ) - 100 )*5-100 }px`;
      else if ( ( clientX - start > 0 && props.left ) || ( clientX - start < 0 && props.right ) )
        component.current.style.left = `${ clientX - start }px` ;

      if ( clientX - start > 60 ) {
        if ( right && right.current )
          right.current.style.opacity = '0';
        if ( left && left.current )
          left.current.style.opacity = `${ ( ( clientX - start ) - 60 )/100 }`;
      } else if ( clientX - start < -60 ) {
        if ( left && left.current )
          left.current.style.opacity = '0';
        if ( right && right.current )
          right.current.style.opacity = `${ -( ( clientX - start ) +60 )/100 }`;
      } else {
        if ( right && right.current )
          right.current.style.opacity = '0';
        if ( left && left.current )
          left.current.style.opacity = '0';
      }
    }

    const onEnd = ( clientX: number, clientY: number ) => {
      const start = +component.current.getAttribute( 'x' );
      const startY = +component.current.getAttribute( 'y' );

      if ( Math.abs( clientY-startY ) < 2 && Math.abs( clientX-start ) < 2 ) props.callback();
      else if ( Math.abs( clientY-startY ) < Math.abs( clientX-start ) && clientX-start >= 100 ) props.left.callback();
      else if ( Math.abs( clientY-startY ) < Math.abs( clientX-start ) && clientX-start <= -100 ) props.right.callback();
      
      reset();
    }

    if ( component && component.current ) {
      component.current.addEventListener( 'mousedown', (e) => onStart( e.clientX, e.clientY ) );
      component.current.addEventListener( 'touchstart', (e) => onStart( e.touches[0].clientX, e.touches[0].clientY ) );
      component.current.addEventListener( 'mousemove', (e) => onMove( e.clientX, e.clientY ) );
      component.current.addEventListener( 'touchmove', (e) => onMove( e.touches[0].clientX, e.touches[0].clientY ) );
      component.current.addEventListener( 'mouseup', (e) => onEnd( e.clientX, e.clientY ) );
      component.current.addEventListener( 'touchend', (e) => onEnd( e.changedTouches[0].clientX, e.changedTouches[0].clientY ) );
      component.current.addEventListener( 'mouseleave', (e) => reset() );
      component.current.addEventListener( 'touchleave', (e) => reset() );
    }
});


  return (
    <Wrapper>
      { props.left ? <Left ref={ left }><Label text={ props.left.label }/></Left> : null }
      <Component ref={ component }>{ props.children }</Component>
      { props.right ? <Right ref={ right }><Label text={ props.right.label }/></Right> : null }
    </Wrapper>
  )
};

export default ActionWrapper;