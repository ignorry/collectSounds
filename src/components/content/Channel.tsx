import React from "react";
import styled from "styled-components";
import { Channel as ChannelType } from "../../models/content";

import Label from "../primitives/Label";

const Container = styled.button`
  color: ${ ({ theme }) => theme.colors.font };
  display: flex;
  align-items: center;
  gap: ${ ({ theme }) => `${ theme.gaps.big }rem` };
  border: none;
`;

const Thumbnail = styled.div`
  height: ${ ({ theme }) => `${ theme.thumbnail.size }rem` };
  position: relative;
`;

const Img = styled.img`
  height: ${ ({ theme }) => `${ theme.thumbnail.size }rem` };
  border-radius: 50%;
  overflow: hidden;
`;

const Info = styled.div`
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  text-align: left;
`;

const TitleContainer = styled.div`
  max-height: ${ ({ theme }) => `${ theme.font.primary.size*2.2 }rem` };
  overflow: hidden;
`;

export type Props = {
  channel: ChannelType;
  callback: Function;
};

const Video: React.FC<Props> = ( props: Props ) => (
  <Container onClick={ () => props.callback() }>
    <Thumbnail>
      <Img
        src={ props.channel.thumbnails.standard?.url || props.channel.thumbnails.default?.url || '' }
      />
    </Thumbnail>

    <Info>
      <TitleContainer>
        <Label text={ props.channel.title } />
      </TitleContainer>
    </Info>
  </Container>
);

export default Video;