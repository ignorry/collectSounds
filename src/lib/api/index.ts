import { 
  fetchVideo,
  fetchPlaylist,
  fetchPlaylistItems,
  fetchChannel,
  fetchSearch
} from "./fetch";
import {
  parseVideo,
  parsePlaylist,
  parseChannel
} from "./parseApiResponse";

import { Video, Playlist, Channel, Content } from "../../models/content";

/**
 * fetch video by id
 * @param {string} id - video id
 * @returns {Promise<Array<Video>>} 
 */
export const getVideoById = async ( id: string ): Promise<Array<Video>> => {
  const res = await fetchVideo({
    part: 'snippet,contentDetails',
    id
  });

  return parseVideo( res );
}

/**
 * fetch playlist by id
 * @param {string} id - playlist id
 * @returns {Promise<Array<Playlist>>}
 */
export const getPlaylistById = async ( id: string ): Promise<Array<Playlist>> => {
  const res = await fetchPlaylist({
    part: 'snippet',
    id
  });

  return parsePlaylist( res );
}

/**
 * fetch channel by id
 * @param {string} id - channel id
 * @returns {Promise<Array<Channel>>}
 */
export const getChannelById = async ( id: string ): Promise<Array<Channel>> => {
  const res = await fetchChannel({
    part: 'snippet',
    id
  });

  return parseChannel( res );
}

/**
 * fetch playlist's videos
 * @param {string} id - playlist id
 * @returns {Promise<Array<Video>>}
 */
export const getPlaylistVideos = async ( id: string ): Promise<Array<Video>> => {
  const res = await fetchPlaylistItems({
    part: 'id',
    playlistId: id,
  });

  const ids: Array<string> = res.items.map( item => item.id );

  return getVideoById( ids.join( ',' ) );
}

/**
 * fetch channel's videos
 * @param {string} id - channel id
 * @returns {Promise<Array<Video>>}
 */
export const getChannelVideos = async ( id: string ): Promise<Array<Video>> => {
  const res = await fetchSearch({
    part: 'id',
    channelId: id,
    order: 'date',
  })

  const ids: Array<string> = res.items.filter( item => item.id.kind === 'video' ).map( item => item.id.videoId );

  return getVideoById( ids.join( ',' ) );
}

/**
 * fetch channel's playlists
 * @param {string} id - channel id
 * @returns {Promise<Array<Playlist>>}
 */
export const getChannelPlaylists = async ( id: string ): Promise<Array<Playlist>> => {
  const res = await fetchPlaylist({
    part: 'snippet',
    channelId: id,
  });

  return parsePlaylist( res )
}

/**
 * @typedef SearchFilters
 * @prop {string} order - specifies the method that will be used to order resources in the API response
 * @prop {string} type - restricts a search query to only retrieve a particular type of resource
 * @prop {string} videoDuration - filters video search results based on their duration
 * @prop {string} saveSearch - indicates whether the search results should include restricted content as well as standard content
 */
export type SearchFilters = {
  order?: 'date' | 'rating' | 'relevance' | 'title' | 'videoCount' | 'viewCount';
  type?: 'video' | 'playlist' | 'channel',
  videoDuration?: 'any' | 'long' | 'medium' | 'short';
  saveSearch?: 'moderate' | 'none' | 'strict'
}

/**
 * fetch search results with specified filters
 * @param {SearchFilters} filters
 * @returns {Promise<Array<Content>>}
 */
export const getSearch = async ( filters: SearchFilters ): Promise<Array<Content>> => {
  const res = await fetchSearch({...filters,
    part: 'id',
    maxResults: 10,
  });

  const videoIds: Array<string> = [];
  const playlistIds: Array<string> = [];
  const channelIds: Array<string> = [];
  const ids: Array<string> = res.items.map( item => {
    if ( item.id.kind === 'video' ) {
      videoIds.push( item.id.videoId );
      return item.id.videoId;
    } else if ( item.id.kind === 'playlist' ) {
      playlistIds.push( item.id.playlistId );
      return item.id.playlistId;
    }

    channelIds.push( item.id.channelId );
    return item.id.channelId;
  });

  const videos: Array<Video> = await getVideoById( videoIds.join( ',' ) );
  const playlists: Array<Playlist> = await getPlaylistById( playlistIds.join( ',' ) );
  const channels: Array<Channel> = await  getChannelById( channelIds.join( ',' ) );

  // ids array needed to save order of contents

  return ids.map( item => {
    if ( videos[0] && videos[0].id === item ) return videos.shift();
    if ( playlists[0] && playlists[0].id === item ) return playlists.shift();
    return channels.shift();
  });
}