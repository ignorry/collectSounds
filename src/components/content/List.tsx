import React, { useRef, useState } from "react";
import { useDispatch } from "react-redux";
import styled from "styled-components";
import { Content } from "../../models/content";
import { useIntl } from "react-intl";

import Label from "../primitives/Label";
import Video from "./Video";
import Playlist from "./Playlist";
import Channel from "./Channel";
import ActionWrapper from "./ActionWrapper";
import { addItem, deleteItem } from "../../redux/slices/saved";
import { addItem as addItemToQueue } from "../../redux/slices/queue";

const Div = styled.div`
  display: flex;
  flex-direction: column;
  gap: ${ ({ theme }) => `${ theme.gaps.big }rem` };
`;

const Message = styled.div`
  display: flex;
  justify-content: center;
  padding: ${ ({ theme }) => `${ theme.gaps.big }rem` };
`;

export type Item = {
  content: Content;
  callback: Function;
}

/**
 * @typedef Props
 * @prop {Array<Item>} items - items to be displayed
 * @prop {string} emptyMessage - display if itmes empty
 */
export type Props = {
  items: Array<Item>;
  emptyMessage: string;
}

const List: React.FC<Props> = ( props: Props ) => {
  const intl = useIntl();
  const save = intl.formatMessage({ id: 'save' }) || '';
  const addToQueue = intl.formatMessage({ id: 'addToQueue' }) || '';
  const del = intl.formatMessage({ id: 'delete' }) || '';

  const dispatch = useDispatch();

  const [hidden, setHidden] = useState<Array<string>>( [] );

  const addHidden = ( id: string ) => setHidden( [...hidden, id] );

  const content: Array<JSX.Element> =
    props.items.map( item => hidden.includes( item.content.id ) ? undefined :
      window.location.pathname.split( '/' )[1] === 'search' ? 
        <ActionWrapper callback={ item.callback } left={{ label: save, callback: () => dispatch( addItem([ item.content ]) ) }}>
          { item.content.type === 'video' ?
            <Video video={ item.content } callback={() => {}}/>
          : item.content.type === 'playlist' ?
            <Playlist playlist={ item.content } callback={() => {}}/>
          : <Channel channel={ item.content } callback={() => {}}/> }
        </ActionWrapper>
      :
        <ActionWrapper callback={ item.callback } 
          left={{ label: addToQueue, callback: () => dispatch( addItemToQueue( item.content.id ) ) }}
          right={{ label: del, callback: () => { dispatch( deleteItem([ item.content.id ]) ); addHidden( item.content.id ) }}}
        >
          { item.content.type === 'video' ?
            <Video video={ item.content } callback={() => {}}/>
          : item.content.type === 'playlist' ?
            <Playlist playlist={ item.content } callback={() => {}}/>
          : <Channel channel={ item.content } callback={() => {}}/> }
        </ActionWrapper>
    );
  
  return (
    <Div>
      {content.length !== 0 ?
        content :
        <Message>
          <Label text={ props.emptyMessage }/>
        </Message>
      }
    </Div>
  );
}

export default List;