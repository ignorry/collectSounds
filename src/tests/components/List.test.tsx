import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { ThemeProvider } from "styled-components";
import { getTheme } from "../../lib/theme";
import GlobalStyles from "../../lib/styled/global";
import { Thumbnails } from "../../models/api/primitives";
import { Video, Playlist, Channel } from "../../models/content";

import List, { Props as ListProps } from "../../components/content/List";

describe( 'List component', () => {
  const prop: ListProps = {
    items: [],
    emptyMessage: 'empty message',
  }

  beforeEach( () => {
    prop.items = [{
      content: {
        type: 'video',
        id: 'id',
        publishedAt: 'publishedAt',
        title: 'title',
        channelTitle: 'channelTitle',
        thumbnails: {
          standard: {
            url: 'http://video.com'
          },
        },
        duration: 59000,
        passed: 5000
      } as Video,
      callback: () => {},
    },{
      content: {
        type: 'playlist',
        id: 'id',
        videos: new Map(),
        title: 'title',
        channelTitle: 'channelTitle',
        thumbnails: {
          standard: {
            url: 'http://playlist.com'
          },
        },
      } as Playlist,
      callback: () => {},
    },{
      content: {
        type: 'channel',
        id: 'id',
        title: 'title',
        thumbnails: {
          standard: {
            url: 'http://channel.com'
          },
        },
      } as Channel,
      callback: () => {},
    }];
  });

  it( 'displays all items', () => {
    const result = render(
      <ThemeProvider theme={ getTheme( true ) } >
        <List
          items={ prop.items }
          emptyMessage={ prop.emptyMessage }
        />
        <GlobalStyles/>
      </ThemeProvider>
    );

    const images = Array.from( result.container.querySelectorAll( 'img' ) ).map( item => item.src );
  
    expect( images ).toStrictEqual([ 'http://video.com/', 'http://playlist.com/', 'http://channel.com/' ]);
  });

  it( 'displays error message', () => {
    prop.items = [];

    const result = render(
      <ThemeProvider theme={ getTheme( true ) } >
        <List
          items={ prop.items }
          emptyMessage={ prop.emptyMessage }
        />
        <GlobalStyles/>
      </ThemeProvider>
    );

    const label = result.container.querySelector( 'p' );
  
    expect( label.innerHTML ).toStrictEqual( 'empty message' );
  });

  it( 'matches to snapshot', () => {
    const result = render(
      <ThemeProvider theme={ getTheme( true ) } >
        <List
          items={ prop.items }
          emptyMessage={ prop.emptyMessage }
        />
        <GlobalStyles/>
      </ThemeProvider>
    );

    expect( result ).toMatchSnapshot();
  });
});