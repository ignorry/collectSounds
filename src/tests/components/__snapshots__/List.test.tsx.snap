// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component matches to snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-hHLeRK iHqIYi"
      >
        <button
          class="sc-dkzDqf jAeszh"
        >
          <div
            class="sc-hKMtZM kJYxUW"
          >
            <img
              class="sc-eCYdqJ eoOQBa"
              src="http://video.com"
            />
            <div
              class="sc-jSMfEi grMTtl"
            >
              <input
                class="sc-gsnTZi loLAgq"
                max="59000"
                type="range"
                value="5000"
              />
            </div>
            <span
              class="sc-gKXOVf gxZVoN"
            >
              <p
                class="sc-bczRLJ gFcYdr"
              >
                59
              </p>
            </span>
          </div>
          <div
            class="sc-iBkjds fGvrQA"
          >
            <div
              class="sc-ftvSup gNwYiK"
            >
              <p
                class="sc-bczRLJ dSaBFM"
              >
                title
              </p>
            </div>
            <div
              class="sc-papXJ zWNMs"
            >
              <p
                class="sc-bczRLJ gFcYdr"
              >
                channelTitle
              </p>
            </div>
            <p
              class="sc-bczRLJ gFcYdr"
            >
              publishedAt
            </p>
          </div>
        </button>
        <button
          class="sc-kDDrLX rAkeG"
        >
          <div
            class="sc-iqcoie AKIOT"
          >
            <img
              class="sc-crXcEl khVjvR"
              src="http://playlist.com"
            />
            <span
              class="sc-evZas bnzzEm"
            >
              <p
                class="sc-bczRLJ dSaBFM"
              >
                0
              </p>
              <svg
                class="sc-jqUVSM gEzVvj"
              >
                <use
                  href="icons.svg#playlistThumbnail"
                />
              </svg>
            </span>
          </div>
          <div
            class="sc-breuTD ojFJZ"
          >
            <div
              class="sc-ksZaOG dROsiD"
            >
              <p
                class="sc-bczRLJ dSaBFM"
              >
                title
              </p>
            </div>
            <div
              class="sc-hAZoDl edibgv"
            >
              <p
                class="sc-bczRLJ gFcYdr"
              >
                channelTitle
              </p>
            </div>
          </div>
        </button>
        <button
          class="sc-fnykZs fQbyBm"
        >
          <div
            class="sc-fEOsli figcqp"
          >
            <img
              class="sc-bjUoiL bveUKH"
              src="http://channel.com"
            />
          </div>
          <div
            class="sc-idiyUo bxcaai"
          >
            <div
              class="sc-dIouRR kgSaMP"
            >
              <p
                class="sc-bczRLJ dSaBFM"
              >
                title
              </p>
            </div>
          </div>
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-hHLeRK iHqIYi"
    >
      <button
        class="sc-dkzDqf jAeszh"
      >
        <div
          class="sc-hKMtZM kJYxUW"
        >
          <img
            class="sc-eCYdqJ eoOQBa"
            src="http://video.com"
          />
          <div
            class="sc-jSMfEi grMTtl"
          >
            <input
              class="sc-gsnTZi loLAgq"
              max="59000"
              type="range"
              value="5000"
            />
          </div>
          <span
            class="sc-gKXOVf gxZVoN"
          >
            <p
              class="sc-bczRLJ gFcYdr"
            >
              59
            </p>
          </span>
        </div>
        <div
          class="sc-iBkjds fGvrQA"
        >
          <div
            class="sc-ftvSup gNwYiK"
          >
            <p
              class="sc-bczRLJ dSaBFM"
            >
              title
            </p>
          </div>
          <div
            class="sc-papXJ zWNMs"
          >
            <p
              class="sc-bczRLJ gFcYdr"
            >
              channelTitle
            </p>
          </div>
          <p
            class="sc-bczRLJ gFcYdr"
          >
            publishedAt
          </p>
        </div>
      </button>
      <button
        class="sc-kDDrLX rAkeG"
      >
        <div
          class="sc-iqcoie AKIOT"
        >
          <img
            class="sc-crXcEl khVjvR"
            src="http://playlist.com"
          />
          <span
            class="sc-evZas bnzzEm"
          >
            <p
              class="sc-bczRLJ dSaBFM"
            >
              0
            </p>
            <svg
              class="sc-jqUVSM gEzVvj"
            >
              <use
                href="icons.svg#playlistThumbnail"
              />
            </svg>
          </span>
        </div>
        <div
          class="sc-breuTD ojFJZ"
        >
          <div
            class="sc-ksZaOG dROsiD"
          >
            <p
              class="sc-bczRLJ dSaBFM"
            >
              title
            </p>
          </div>
          <div
            class="sc-hAZoDl edibgv"
          >
            <p
              class="sc-bczRLJ gFcYdr"
            >
              channelTitle
            </p>
          </div>
        </div>
      </button>
      <button
        class="sc-fnykZs fQbyBm"
      >
        <div
          class="sc-fEOsli figcqp"
        >
          <img
            class="sc-bjUoiL bveUKH"
            src="http://channel.com"
          />
        </div>
        <div
          class="sc-idiyUo bxcaai"
        >
          <div
            class="sc-dIouRR kgSaMP"
          >
            <p
              class="sc-bczRLJ dSaBFM"
            >
              title
            </p>
          </div>
        </div>
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
